### PR TITLE
Coexistance of cellfinder+brainreg CLI tool and cellfinder-core workflow

### PR DESCRIPTION
**Pull request by sfmig**
### Context
This repository currently serves a few purposes:
- Host the `cellfinder` CLI tool (onwards called '`cellfinder+brainreg` CLI tool' for clarity).
- Host example workflow scripts. 
    - These are useful for developers to benchmark (see next point) and for users only as scripts to "copy-paste" and adapt to their own use-cases. Right now there is only one`cellfinder-core` workflow script.
- Host benchmarks that time the workflow scripts. 
    - These are only useful for developers.

Our understanding is that the `cellfinder+brainreg` CLI tool will be gradually discontinued (in line with the `brainglobe` general aim of removing CLI tools) and turned into a workflow script instead. This PR outlines a strategy to get there.

### Goal of this PR
With this PR, the `brainglobe-workflows` PyPI package will ship:
 - the `cellfinder+brainreg` CLI tool, and 
 - the `cellfinder_core` workflow script.

The `cellfinder_core` workflow script is only included in the source code,  and users are not meant to be "aware" of it (aka, it is not documented in the user docs). Users will use the `cellfinder+brainreg` CLI tool in the same way as they did before.

Instead, developers _are_ aware of the `cellfinder_core` workflow script and of the benchmarks defined on top of it (that is, both things are documented in the dev docs).

#### How do users use this repo?
Users will `pip install brainglobe-workflows` and use the `cellfinder+brainreg` CLI tool as usual, with the `cellfinder` entrypoint.

#### How do devs use this repo?
Devs will:
1. git clone this repo
2. `pip install .[dev]`
3. Run benchmarks locally with `asv run` or run tests with `tox`/`pytest`

#### Three sets of dependencies
To cater for these different audiences, we divided the dependencies in `pyproject.toml` into three sets:
- `dependencies`, the required ones. We propose that these contain the dependencies for the `cellfinder+brainreg` CLI tool _only_, for now (see [Longer term](#longer-term) section).
- `dev`, contains the dependencies required by developers to contribute to the repo (run tests, benchmarks, etc). 
- `asv_version`, contains the dependencies required by `asv` to be able to run the benchmarks. 
    - The idea is that these are the dependencies needed to install the package _without_ the CLI tool. 
    - Once the `cellfinder+brainreg` CLI tool is deprecated, these dependencies will move to the default dependencies section.
    - Note that asv  works by creating a new virtual env, installing the software version of a given commit into it, and running the benchmarking functions on it.

### Longer term
- In the short-medium term, we need to ensure PRs don't break the `cellfinder+brainreg` CLI tool.
- The idea would be to only update the package in PyPI when the CLI tool is updated.
- In the longer term, the `cellfinder+brainreg` CLI tool will be transformed into a workflow script, and we will no longer have the need to publish a package in PyPI of this repo. When that happens all of the dependencies under `asv_version` would move to the required dependencies section.

### Specifics of this PR
- Restructured package internals (and adapted entry point accordingly)
- Modified the asv package install command, to install the brainglobe-workflows package with `asv_version` dependencies only.
- Updated the README.md to split user and dev goals more clearly.
- Defined three sets of dependencies in pyproject.toml

Rebase after merging #27

### CODE REVIEW

**Comment by adamltyson** [comment]
I might have missed something, but is this needed? What used to be called `cellfinder-core` is now called `cellfinder` anyway now.

**Comment by alessandrofelder** [comment]
We need to somehow distinguish the CLI tool (present in this repo) from benchmarked scripts using the Python tool's API?

**Comment by sfmig** [comment]
my understanding is that the cellfinder+brainreg CLI tool is called in this repo `cellfinder` after the migration of the cellfinder repo to the brainglobe-workflows repo. This is in conflict with me previously using `cellfinder` to define a cellfinder-core workflow.

**Comment by adamltyson** [comment]
> my understanding is that the cellfinder+brainreg CLI tool is called in this repo cellfinder after the migration of the cellfinder repo to the brainglobe-workflows repo. This is in conflict with me previously using cellfinder to define a cellfinder-core workflow.

This [should change soon anyway](https://github.com/brainglobe/brainglobe-workflows/issues/60), but maybe it's good to specify `cellfinder_core` to distinguish from `cellfinder.napari` anyway.

File: README.md [hunk 1]
```diff
@@ -16,16 +16,23 @@
...
-At present, the package offers the following workflows:
+At present, the package offers the following workflows to users:
+
+- [cellfinder](#cellfinder): a command-line tool for whole-brain detection, registration, and analysis.
 
-- [cellfinder](#cellfinder): Whole-brain detection, registration, and analysis.
+Additionally, this repository provides functionalities to support code developers. See [Developers documentation](#developers-documentation) for further details.
 
-## Installation
+## Users documentation
```
Comment (willGraham01) [inline README.md:31]:
```suggestion
## User documentation
```

File: README.md [hunk 2]
```diff
@@ -34,32 +41,16 @@ pip install brainglobe-workflows
...
-If you find [`cellfinder`](#cellfinder) useful, and use it in your research, please cite the paper outlining the cell detection algorithm:
-> Tyson, A. L., Rousseau, C. V., Niedworok, C. J., Keshavarzi, S., Tsitoura, C., Cossell, L., Strom, M. and Margrie, T. W. (2021) “A deep learning algorithm for 3D cell detection in whole mouse brain image datasets’ PLOS Computational Biology, 17(5), e1009074
-[https://doi.org/10.1371/journal.pcbi.1009074](https://doi.org/10.1371/journal.pcbi.1009074)
->
-If you use any of the image registration functions in `cellfinder`, please also cite [`brainreg`](https://github.com/brainglobe/brainreg#citing-brainreg).
-
----
 
-## Cellfinder
+### Cellfinder CLI tool
```
Comment (willGraham01) [inline README.md:45]:
```suggestion
### Cellfinder Command Line Interface (CLI)
```

File: README.md [hunk 3]
```diff
@@ -73,3 +64,27 @@ cellfinder -s signal_images -b background_images -o output_dir --metadata metada
 ```
 
 Full documentation can be found [here](https://brainglobe.info/documentation/cellfinder/index.html).
+
+
+## Developers documentation
```
Comment (willGraham01) [inline README.md:69]:
```suggestion
## Developer documentation
```

File: README.md [hunk 4]
```diff
@@ -73,3 +64,27 @@ cellfinder -s signal_images -b background_images -o output_dir --metadata metada
...
+
+This repository also includes workflow scripts that are benchmarked to support code development. We regularly benchmark the time the workflows take to complete to ensure performance is stable as the code changes.
```
Comment (willGraham01) [inline README.md:71]:
```suggestion
This repository also includes workflow scripts that are benchmarked to support code development.
These benchmarks are run regularly to ensure performance is stable, as the tools are developed and extended.
```

Also a handy tip for markdown files; start every sentence on a newline (if you're not in a list) - it means that hunting through the diff for changes and edits is a little bit faster :slightly_smiling_face:

File: README.md [hunk 5]
```diff
@@ -73,3 +64,27 @@ cellfinder -s signal_images -b background_images -o output_dir --metadata metada
...
+* Developers can install these benchmarks locally via `pip install .[dev]`. By executing `asv run`,  the benchmarks will run with default parameters on a small dataset that is downloaded from [GIN](https://gin.g-node.org/G-Node/info/wiki). See [the asv docs](https://asv.readthedocs.io/en/v0.6.1/using.html#running-benchmarks) for further details on how to run benchmarks.
```
Comment (willGraham01) [inline README.md:72]:
```suggestion
* Developers can install these benchmarks locally via `pip install .[dev]`. By executing `asv run`, the benchmarks will run with default parameters on a small dataset that is downloaded from [GIN](https://gin.g-node.org/G-Node/info/wiki). See [the asv docs](https://asv.readthedocs.io/en/v0.6.1/using.html#running-benchmarks) for further details on how to run benchmarks.
```

(VSCode markdown lint is viscous)

File: README.md [hunk 6]
```diff
@@ -73,3 +64,27 @@ cellfinder -s signal_images -b background_images -o output_dir --metadata metada
...
+* Developers can also run these benchmarks on  data available locally, by specifying the relevant paths in an input configuration JSON file.
```
Comment (willGraham01) [inline README.md:73]:
```suggestion
* Developers can also run these benchmarks on data they have stored locally, by specifying the relevant paths in an input (JSON) file.
```

File: README.md [hunk 7]
```diff
@@ -73,3 +64,27 @@ cellfinder -s signal_images -b background_images -o output_dir --metadata metada
...
+* We additionally run and benchmark the workflows locally on a internal desktop machine using large example datasets. These benchmarks are run periodically and the results are made publicly available.
```
Comment (willGraham01) [inline README.md:74]:
```suggestion
* We also maintain an internal runner that benchmarks the workflows over a large, exemplar dataset, of the scale we expect users to be handling. The result of these benchmarks are made publicly available.
```

File: README.md [hunk 8]
```diff
@@ -73,3 +64,27 @@ cellfinder -s signal_images -b background_images -o output_dir --metadata metada
...
+
+Contributions to BrainGlobe are more than welcome.
+Please see the [developers guide](https://brainglobe.info/developers/index.html).
```
Comment (willGraham01) [inline README.md:77]:
```suggestion
Please see the [developer guide](https://brainglobe.info/developers/index.html).
```

**Review by willGraham01 — COMMENTED** [review]
The plan in the description sounds a lot more synthesised than in our Thursday programming session, great job :smile: 

Some comments on the README but most of them are me just rephrasing things or VSCode markdown lint getting angry. Haven't touched the ASV stuff because it scares me.

**Comment by adamltyson** [comment]
> In the longer term, the cellfinder+brainreg CLI tool will be transformed into a workflow script

Slight comment, I think we will maintain some CLIs (i.e. just the command line interface) that are heavily used. However, I think we should move most of the code inside these workflows to other packages so that it can be a single workflow script.

File: README.md (hunk 4 above)
Comment (sfmig) [reply README.md:71]: I was wondering if there was a pre-commit hook for this but didn't find any already made

Comment (willGraham01) [reply README.md]: There isn't one sadly, but I have my VSCode setup with a markdown linter which helps standardise some things.
https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one

**Review by alessandrofelder — APPROVED** [review]
Great - happy with this plan.

**Pull request merged**